src/sinks/util/unix.rs

```diff
@@ -136,131 +142,58 @@ impl UnixSink {
-                },
-                UnixSinkState::Disconnected => {
-                    debug!(
-                        message = "Connecting",
-                        path = %self.connector.path.to_str().unwrap()
+            debug!(
+                message = "Connecting",
+                path = %self.path.to_str().unwrap()
+            );
+            match UnixStream::connect(self.path.clone()).await {
+                Ok(stream) => {
+                    emit!(UnixSocketConnectionEstablished { path: &self.path });
+                    return AckerBytesSink::new(
+                        stream,
+                        self.acker.clone(),
+                        Box::new(|byte_size| emit!(UnixSocketEventSent { byte_size })),
                     );
-                    let fut = self.connector.connect();
-                    UnixSinkState::Creating(Box::new(fut.compat()))
                 }
-            }
-        }
-    }
-}
-
-impl Sink for UnixSink {
-    type SinkItem = Bytes;
-    type SinkError = ();
-
-    fn start_send(&mut self, line: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
-        let byte_size = line.len();
-        match self.poll_connection() {
-            Ok(Async::Ready(connection)) => match connection.start_send(line) {
                 Err(error) => {
-                    emit!(UnixSocketSendFailed {
+                    emit!(UnixSocketConnectionFailure {
                         error,
-                        path: &self.connector.path,
+                        path: &self.path
                     });
-                    self.state = UnixSinkState::Disconnected;
-                    Ok(AsyncSink::Ready)
-                }
-                Ok(res) => {
-                    emit!(UnixSocketEventSent { byte_size });
-                    Ok(res)
+                    delay_for(backoff.next().unwrap()).await;
                 }
-            },
-            Ok(Async::NotReady) => Ok(AsyncSink::NotReady(line)),
-            Err(_) => unreachable!(),
+            }
         }
     }
+}
 
-    fn poll_complete(&mut self) -> Poll01<(), Self::SinkError> {
-        // Stream::forward will immediately poll_complete the sink it's forwarding to,
-        // but we don't want to connect before the first event actually comes through.
-        if let UnixSinkState::Disconnected = self.state {
-            return Ok(Async::Ready(()));
-        }
-
-        let connection = try_ready!(self.poll_connection());
-
-        match connection.poll_complete() {
-            Err(error) => {
-                emit!(UnixSocketFlushFailed {
+#[async_trait]
+impl StreamSink for UnixSink {
+    async fn run(&mut self, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        let encoding = self.encoding.clone();
+        let mut input = input
+            .map(|event| match encode_event(event, &encoding) {
+                Some(bytes) => bytes,
+                None => Bytes::new(),
```

Comment:
Can we use `filter_map` here to avoid sending an empty `Bytes` to the sink?

Reply:
Yes, we can, but then we will need to add `ack(1)` and extra `emit!`. I just double-checked `FramedWrite` and `BytesCodec` -- it's completely okay send empty `Bytes`, only a few extra calls which do nothing.

Reply:
This seems subtle and can lead to incorrect acking, so I think we should add a comment for it, may be at AckerBytesSink?

Reply:
Yeah, comment would be great, I'll add.